File: android/app/src/main/kotlin/com/danieldallos/pigeon_multi_instance_demo/pigeon/PigeonMultiInstanceBinaryMessengerWrapper.kt
package com.danieldallos.pigeon_multi_instance_demo.pigeon

import io.flutter.plugin.common.BinaryMessenger
import java.nio.ByteBuffer

class PigeonMultiInstanceBinaryMessengerWrapper(
    private val messenger: BinaryMessenger,
    private val channelSuffix: String,
) : BinaryMessenger {

    override fun send(channel: String, message: ByteBuffer?) {
        messenger.send("$channel/$channelSuffix", message)
    }

    override fun send(channel: String, message: ByteBuffer?, callback: BinaryMessenger.BinaryReply?) {
        messenger.send("$channel/$channelSuffix", message, callback)
    }

    override fun setMessageHandler(channel: String, handler: BinaryMessenger.BinaryMessageHandler?) {
        messenger.setMessageHandler("$channel/$channelSuffix", handler)
    }
}

File: ios/Runner/pigeon/PigeonMultiInstanceBinaryMessengerWrapper.swift
//
//  PigeonMultiInstanceBinaryMessengerWrapper.swift
//  Runner
//
//  Created by Daniel on 10/10/2023.
//

import Foundation
import Flutter

class PigeonMultiInstanceBinaryMessengerWrapper: NSObject, FlutterBinaryMessenger {

    let channelSuffix: String
    let messenger: FlutterBinaryMessenger

    init(with messenger: FlutterBinaryMessenger, channelSuffix: String) {
        self.messenger = messenger
        self.channelSuffix = channelSuffix
    }

    func send(onChannel channel: String, message: Data?, binaryReply callback: FlutterBinaryReply? = nil) {
        messenger.send(onChannel: "\(channel)/\(channelSuffix)", message: message, binaryReply: callback)
    }

    func send(onChannel channel: String, message: Data?) {
        messenger.send(onChannel: "\(channel)/\(channelSuffix)", message: message)
    }

    func setMessageHandlerOnChannel(_ channel: String, binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil) -> FlutterBinaryMessengerConnection {
        messenger.setMessageHandlerOnChannel("\(channel)/\(channelSuffix)", binaryMessageHandler: handler)
    }

    func cleanUpConnection(_ connection: FlutterBinaryMessengerConnection) {
        messenger.cleanUpConnection(connection)
    }
}
